MODOAIPMH-300: Not all instances ids are saved for initial harvest

IlliaDaliek opened the pull request:
Jira: https://issues.folio.org/browse/MODOAIPMH-300
Jira: https://issues.folio.org/browse/MODOAIPMH-301

### PURPOSE
Fix issue with the hanging of srs response handler which caused the harvest to be canceled. 
The root cause is that when srs responds with a large response body it may hang for an unknown reason and the body is not written to the response, therefore we can get the body and invoke body handler which causes the code hanging and canceling harvesting as result.

### APPROCH

Set the idle timeout of srs request to 20 seconds (instead of 5000) in order to invoke the exception handler (by causing timeout exception) when there a problem with receiving response body from SRS. Make a retry of srs call in the exception handler.

IlliaDaliek commented on src/test/java/org/folio/rest/impl/OaiPmhImplTest.java:
```diff
@@ -2196,9 +2199,10 @@ void shouldRespondWithInternalServerError_whenGetOaiRecordsMarc21WithHoldingsWit
       .extract()
       .asString();
-    String expectedMessage = format(EXPECTED_ERROR_MESSAGE, "source-record-storage");
+    String expectedMessage = "SRS didn't respond with expected status code after 1 attempts. Canceling further request processing.";
```
Unit test for re-requesting srs was written before this PR
It is called shouldRerequestSrsWhenItRespondedWith500_whenGetOaiRecordsMarc21WithHoldingsWithErrorResponseAndThenSuccessResponseFromSrs

craigmcnally commented on src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:
```diff
@@ -2,38 +2,32 @@
 
+  private static final Logger logger = LoggerFactory.getLogger(SourceStorageSourceRecordsClient.class);
+
   public SourceStorageSourceRecordsClient(String okapiUrl, String tenantId, String token, boolean keepAlive, int connTO, int idleTO) {
```
Since units were a problem here, it may be a good idea to incorporate the units into the variable names... e.g. `idleTimeoutSec`, `connTimeoutMs`

IlliaDaliek replied:
Agree, corrected.

craigmcnally commented on src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:
```diff
@@ -69,8 +67,14 @@ public void postSourceStorageSourceRecords(String idType, Boolean deleted, List
       buffer.appendString(ClientHelpers.pojo2json(List));
     }
 
+
```
```suggestion
```

craigmcnally commented on src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:
```diff
@@ -441,6 +439,7 @@ private void getNextBatch(String requestId, Request request, boolean firstBatch,
       return completePromise.future();
     }
 
+
```
```suggestion
```

craigmcnally replied:
Can we remove this extra newline?

craigmcnally commented on src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:
```diff
@@ -47,10 +41,14 @@ public SourceStorageSourceRecordsClient(String okapiUrl, String tenantId, String
 
   public SourceStorageSourceRecordsClient(String okapiUrl, String tenantId, String token) {
-    this(okapiUrl, tenantId, token, true, 2000, 5000);
+    this(okapiUrl, tenantId, token, true, 2000, 20);
```
Instead of hardcoding these timeout values, can we make them constants?

IlliaDaliek replied:
Yes, done.

craigmcnally commented on src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:
```diff
@@ -69,8 +67,14 @@ public void postSourceStorageSourceRecords(String idType, Boolean deleted, List
     HttpClientRequest request = this.httpClient.postAbs(this.okapiUrl + "/source-storage/source-records" + queryParams.toString());
     request.handler(responseHandler);
+    request.exceptionHandler(e-> {
+      logger.error("SRS response error 1234: " + e.getMessage(), e);
```
what is the significance of error 1234?

craigmcnally replied:
```suggestion
      logger.error("Exception making SRS request: " + e.getMessage(), e);
```

IlliaDaliek replied:
Thanks for the note, initially this log was added just for testing of implemented logic work. I deleted it since we have loggin in the "srs retry method".

craigmcnally commented on src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:
```diff
@@ -69,8 +67,14 @@ public void postSourceStorageSourceRecords(String idType, Boolean deleted, List
+      logger.error("SRS response error 1234: " + e.getMessage(), e);
+      exceptionHandler.handle(e);
+    });
+    request.setTimeout(10000);
```
See comment above about using constants instead of integer literals

IlliaDaliek replied:
Done

craigmcnally commented on src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:
```diff
@@ -694,74 +692,82 @@ private HttpClientRequest createEnrichInventoryClientRequest(HttpClient httpClie
-        srsResponse.bodyHandler(getSrsResponseBodyHandler(srsClient, promise));
-      });
+        srsResponse.bodyHandler(buffer -> handleSrsResponse(srsClient, promise, buffer));
+      }, e->  retrySRSRequest(srsClient, vertx, deletedRecordSupport, listOfIds, attemptsCount, retryAttempts, promise, 400, "Error in SRS response"));
```
I think it's misleading to always report 400 status here...   It's probably better to not report a status than to potentially report the wrong one.

IlliaDaliek replied:
Agree, corrected.

craigmcnally commented on src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:
```diff
@@ -694,74 +692,82 @@ private HttpClientRequest createEnrichInventoryClientRequest(HttpClient httpClie
-        srsResponse.bodyHandler(getSrsResponseBodyHandler(srsClient, promise));
-      });
+        srsResponse.bodyHandler(buffer -> handleSrsResponse(srsClient, promise, buffer));
+      }, e->  retrySRSRequest(srsClient, vertx, deletedRecordSupport, listOfIds, attemptsCount, retryAttempts, promise, 400, "Error in SRS response"));
```
If I read this right, the exception handler provided here will only be invoked upon request exceptions (`request.exceptionHandler(...)`).  In this case there is no status code.  Furthermore, the message provided `Error in SRS response` is inaccurate.  `Error making SRS request` would be more accurate.

craigmcnally commented on src/main/java/org/folio/oaipmh/processors/MarcWithHoldingsRequestHelper.java:
```diff
@@ -694,74 +692,82 @@ private HttpClientRequest createEnrichInventoryClientRequest(HttpClient httpClie
-      } finally {
-        srsClient.close();
+  private void retrySRSRequest(SourceStorageSourceRecordsClient srsClient, Vertx vertx, boolean deletedRecordSupport, List<String> listOfIds, AtomicInteger attemptsCount, int retryAttempts, Promise<Map<String, JsonObject>> promise, int statusCode, String statusMessage) {
+    String warnMessage = "Got error response form SRS, status code: " + statusCode + ", status message: " + statusMessage;
+    logger.debug(warnMessage);
```
It's probably better to log this before calling this method.  The way this currently stands, in some cases you might log a message at ERROR level, then a different (also inaccurate - see earlier comments) message at WARN level.

craigmcnally replied:
I still see you passing a hard-coded 400 into this method in the case where there is no response code - because the error occurred while writing the request.  Changing the log level from WARN to DEBUG is not sufficient to fix this IMO.

IlliaDaliek replied:
> I still see you passing a hard-coded 400 into this method in the case where there is no response code - because the error occurred while writing the request. Changing the log level from WARN to DEBUG is not sufficient to fix this IMO.
@craigmcnally, it's not correct. 
There re-request can be triggered by two reasons: 
1) when the srs returns a full response but with incorrect status code
2) when there a problem in srs with passing the body and the timeout exception is thrown (or any else potential http exceptions).
These two cases are different situations and they should have different log messages. I corrected it and we didn't pass the hardcoded "400" value. 
![image](https://user-images.githubusercontent.com/60663044/110809886-f6f23480-828d-11eb-9c6f-a7988d100290.png)

craigmcnally commented on src/main/java/org/folio/oaipmh/Constants.java:
```diff
@@ -25,6 +23,7 @@ private Constants() {
 
   public static final String REPOSITORY_BASE_URL = "repository.baseURL";
   public static final String REPOSITORY_MAX_RECORDS_PER_RESPONSE = "repository.maxRecordsPerResponse";
+  public static final String REPOSITORY_HTTP_REQUEST_RETRY_ATTEMPTS = "repository.httpRequestRetryAttempts";
```
`httpRequestRetryAttempts` is a little misleading - The module makes several HTTP requests.  This configuration only applies to the call to SRS.

IlliaDaliek replied:
Okay, it makes sense. I added the "srs" prefix to the property name.

craigmcnally commented on pom.xml:
```diff
@@ -76,12 +76,12 @@
         <groupId>io.vertx</groupId>
         <artifactId>vertx-sql-client</artifactId>
-        <version>${vertx.version}-FOLIO</version>
+        <version>3.9.4-FOLIO</version>
```
I didn't see this mentioned in the PR description... is this a necessary part of this PR?  If so, can you please explain why?

IlliaDaliek replied:
When I had a discussion with Adam and Julian they suggested to use the latest vertx version with major version 3 (3.9.6).
But dependencies with the FOLIO prefix for 3.9.6 don't exist. So we changed the vertx version property value to be 3.9.6 and left some dependencies with 3.9.4-FOLIO.

craigmcnally replied:
In the future, please add this context to the PR description so that it's clear why the change is being made.  Without this information the PR reviewer is left wondering if the change was intentional, or a mistake.

MikhailFokanov reviewed (approved):
Looks good.

craigmcnally commented on src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:
```diff
@@ -2,55 +2,58 @@
-  private String tenantId;
-  private String token;
-  private String okapiUrl;
+  private static final int DEFAULT_SRS_TIMEOUT = 10000;
```
why no units on this one?

IlliaDaliek replied:
I added a test.

craigmcnally commented on src/main/java/org/folio/oaipmh/client/SourceStorageSourceRecordsClient.java:
```diff
@@ -69,8 +72,10 @@ public void postSourceStorageSourceRecords(String idType, Boolean deleted, List
     }
 
-    HttpClientRequest request = this.httpClient.postAbs(this.okapiUrl + "/source-storage/source-records" + queryParams.toString());
+    HttpClientRequest request = this.httpClient.postAbs(this.okapiUrl + GLOBAL_PATH + queryParams.toString());
```
I realize this was already there and you're just using it but `GLOBAL_PATH` is a lousy name for this.  Something like `SOURCE_RECORDS_PATH` is IMO much more appropriate

craigmcnally reviewed (dismissed):
A couple minor comments related to variable names.  Approving now, but please take my comments into consideration.

craigmcnally reviewed (changes requested):
Please address the issues raised about the logging in retry method

craigmcnally commented:
@IlliaDaliek I noticed that you've introduce new changes here after the PR was approved.  Please re-request reviews from those that previously approved.  This feels a little like scope creep here.  Strictly speaking this probably should have happened in a separate PR, probably with its own JIRA.  Given that the changes are already here, we might as well go with it.

IlliaDaliek commented:
> @IlliaDaliek I noticed that you've introduce new changes here after the PR was approved. Please re-request reviews from those that previously approved. This feels a little like scope creep here. Strictly speaking this probably should have happened in a separate PR, probably with its own JIRA. Given that the changes are already here, we might as well go with it.

@craigmcnally, there no serious changes, I moved srs idle timeout to module settings as Mikhail asked me to do that. As well when I verified the "schema update testing" ticket I noticed incorrect logs and fixed them as well in the current PR. And among the code the laterals were used instead of already defined constants, I corrected this too.

craigmcnally commented:
> > @IlliaDaliek I noticed that you've introduce new changes here after the PR was approved. Please re-request reviews from those that previously approved. This feels a little like scope creep here. Strictly speaking this probably should have happened in a separate PR, probably with its own JIRA. Given that the changes are already here, we might as well go with it.
> 
> @craigmcnally, there no serious changes, I moved srs idle timeout to module settings as Mikhail asked me to do that. As well when I verified the "schema update testing" ticket I noticed incorrect logs and fixed them as well in the current PR. And among the code the laterals were used instead of already defined constants, I corrected this too.

OK thanks.  I can't stress this enough how important it is to communicate these details in the PR.  Otherwise, it's very hard to understand what's going on, and why.
